### PR TITLE
Add retry_if_not_exception_type()

### LINKED ISSUE #256: Retry if not exception type?
Hi, I was wondering if there's an easy way to say retry for everything except this list of exceptions?

For example, I want to retry for all server errors but not client errors.

### PR DESCRIPTION
**Pull request by henry0312**
Fixes #256

### CODE REVIEW

File: releasenotes/notes/add-retry_except_exception_type-31b31da1924d55f4.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+features:
+  - Add ``retry_except_exception_type()`` that allows us to retry if a raised exception doesn't match given exceptions.
```
Comment (jd) [inline releasenotes/notes/add-retry_except_exception_type-31b31da1924d55f4.yaml:3]:
```suggestion
  - Add ``retry_except_exception_type()`` that allows to retry if a raised exception doesn't match given exceptions.
```

File: tenacity/retry.py [hunk 2]
```diff
@@ -80,6 +80,16 @@ def __init__(self, exception_types=Exception):
         )
 
 
+class retry_except_exception_type(retry_if_exception):
+    """Retries excpet an exception has been raised of one or more types."""
```
Comment (jd) [inline tenacity/retry.py:84]:
```suggestion
    """Retries except an exception has been raised of one or more types."""
```

**Review by jd — CHANGES_REQUESTED** [review]
Small typo

**Review by jd — DISMISSED** [review]

**Comment by asqui** [comment]
I think we can be clearer on the naming of this method to better communicate what it is doing.

Is this implementing the `retry_if_exception_not_of_type` which was propossed in the discussion of #137?

**Comment by henry0312** [comment]
> Is this implementing the retry_if_exception_not_of_type which was propossed in the discussion of #137?

Yes, it it.
Though @jd approved this, would it be good to rename `retry_except_exception_type` to `retry_if_exception_not_of_type`?

**Comment by jd** [comment]
I'm fine with whatever.
We need to fix the CI not running issue to merge anything at this point 😢

**Comment by asqui** [comment]
I do think `retry_if_exception_not_of_type` would fit better amongst the existing family of related `retry_*` methods:

- `retry_if_exception`
- `retry_if_exception_type`
- `retry_unless_exception_type` -- would have been better named `retry_if_not_exception_type`, but yolo.
- `retry_if_result`
- `retry_if_not_result`
- `retry_if_exception_message`
- `retry_if_not_exception_message`

In particular, I think it's important that the new method, whatever it's called, is clearly set apart from `retry_unless_exception_type` given the subtle but important differences between the two (as discussed, at length, in #137).

More generally, I'm slightly alarmed by the "I'm fine with whatever" remark. I worry that without careful consideration of usability, adoption of this library will be hampered in the future. (This is speaking as someone who has tried to bring tenacity in to an existing codebase which already has numerous disparate approaches to retry logic, and the complexity/usability concerns around tenacity have certainly been a barrier to adoption.)

**Comment by henry0312** [comment]
A relation between `retry_if_exception_type()` and `retry_except_exception_type()` is the same with one between `retry_if_exception_message()` and `retry_if_not_exception_message()`, since the implementation of `retry_except_exception_type()` is the essentially same with that of `retry_if_not_exception_message()`.
(cf. https://github.com/jd/tenacity/blob/master/tenacity/retry.py#L125 and https://github.com/jd/tenacity/blob/master/tenacity/retry.py#L167)

Therefore, if I would rename, I think `retry_if_not_exception_type` is better than `retry_if_exception_not_of_type`, that follows the named rule of `retry_if_not_exception_message`.

---
For details,

- `retry_if_exception_type(T)`

`T` is **included** in conditons of retry.

- `retry_except_exception_type(T)`

`T` is **excluded** from conditions of retry.
In other words, `T` is **not included** in condtions of retry.

- `retry_if_exception_message(A)`

`A` is **included** in conditons of retry.

- `retry_if_not_exception_message(A)`

`A` is **excluded** from conditions of retry.
This means that `T` is **not included** in condtions of retry.

By the way,

- `retry_unless_exception_type(T)`

Retry **until** `T` is raised.
(`T` is expected to be raised to stop retry)

---

I'm not good ad English and I'm sorry for my bad English.

**Comment by asqui** [comment]
@henry0312 Thanks for taking the time to look at this in such detail.

I agree that `retry_if_not_exception_type` fits better with the existing naming conventions.

P.S. Don't apologise, your english is great! (I have no reason to believe you're not a native speaker!)

**Comment by henry0312** [comment]
@jd @asqui I just renamed `retry_except_exception_type` to `retry_if_not_exception_type`

**Review by jd — APPROVED** [review]

**Comment by henry0312** [comment]
@jd I just fixed tests!

**Review by jd — APPROVED** [review]

**Comment by henry0312** [comment]
@jd Thank you for reviewing and merging.
Do you have a plan for a new [release](https://github.com/jd/tenacity/releases)?

**Comment by william-silversmith** [comment]
Likewise! Thanks so much jd and all that contributed! I would start using this feature once the next release is out.
